### PR TITLE
Verify graders Take 3

### PR DESCRIPTION
**Pull request by jchulton**
The last Verify graders reverted 3000 lines in its second commit, so I decided just to make a new pull request on a new branch

### CODE REVIEW

File: migration/data/course_tables.sql [hunk 1]
```diff
@@ -540,29 +541,29 @@ CREATE TABLE forum_posts_history (
 );
 
 CREATE TABLE "thread_categories" (
-	"thread_id" int NOT NULL,
-	"category_id" int NOT NULL
+    "thread_id" int NOT NULL,
```
Comment (KevinMackenzie) [inline migration/data/course_tables.sql]: Probably should remove these whitespace changes since they aren't directly related to the PR

File: migration/migrations/course/20180611120540_regrade_request.py [hunk 2]
```diff
@@ -1,5 +1,6 @@
 def up(config, conn, semester, course):
     with conn.cursor() as cursor:
+
```
Comment (KevinMackenzie) [inline migration/migrations/course/20180611120540_regrade_request.py]: Remove whitespace change. Migration files should almost never be modified after their initial creation

File: site/app/controllers/grading/ElectronicGraderController.php [hunk 3]
```diff
@@ -1041,10 +1041,14 @@ public function saveSingleComponent() {
             $component->deleteData($gradeable->getGdId());
             $debug = 'delete';
         } else {
-            //only change the component information is the mark was modified or componet and its gradeable are out of sync.
-            if ($component->getGrader() === null || $overwrite === "true") {
-                $component->setGrader($this->core->getUser());
-            }
+                //only change the component information is the mark was modified or componet and its gradeable are out of sync.
```
Comment (KevinMackenzie) [inline site/app/controllers/grading/ElectronicGraderController.php]: '... component information *if* the mark was modified or *component* and ...' I know they aren't your spelling errors, but since you're modifying the line anyway can you fix them?

File: site/app/libraries/database/DatabaseQueries.php [hunk 4]
```diff
@@ -1306,14 +1314,14 @@ public function insertGradeableComponentData($gd_id, GradeableComponent $compone
      * @param string             $grader_id
      * @param GradeableComponent $component
      */
-    public function updateGradeableComponentData($gd_id, $grader_id, GradeableComponent $component) {
+    public function updateGradeableComponentData($gd_id, $grader_id, $grader2_id, GradeableComponent $component) {
```
Comment (KevinMackenzie) [inline site/app/libraries/database/DatabaseQueries.php]: `verifier_id`

File: site/app/libraries/database/DatabaseQueries.php [hunk 5]
```diff
@@ -283,8 +283,8 @@ private function findChildren($post_id, $thread_id, &$children, $get_deleted = f
     }
 
     public function searchThreads($searchQuery){
-    	$this->course_db->query("SELECT post_content, p_id, p_author, thread_id, thread_title, author, pin, anonymous, timestamp_post FROM (SELECT t.id as thread_id, t.title as thread_title, p.id as p_id, t.created_by as author, t.pinned as pin, p.timestamp as timestamp_post, p.content as post_content, p.anonymous, p.author_user_id as p_author, to_tsvector(p.content) || to_tsvector(p.author_user_id) || to_tsvector(t.title) as document from posts p, threads t JOIN (SELECT thread_id, timestamp from posts where parent_id = -1) p2 ON p2.thread_id = t.id where t.id = p.thread_id and p.deleted=false and t.deleted=false) p_doc JOIN (SELECT thread_id as t_id, timestamp from posts where parent_id = -1) p2 ON p2.t_id = p_doc.thread_id  where p_doc.document @@ plainto_tsquery(:q)", array(':q' => $searchQuery));
-    	return $this->course_db->rows();
```
Comment (KevinMackenzie) [inline site/app/libraries/database/DatabaseQueries.php]: more unrelated whitespace changes

File: site/app/libraries/database/PostgresqlDatabaseQueries.php [hunk 6]
```diff
@@ -1288,14 +1314,14 @@ public function getGradedGradeables(array $gradeables, $users = null, $teams = n
...
                 SELECT
-                  json_agg(in_gcd.gc_id) AS array_comp_id,
-                  json_agg(gcd_score) AS array_score,
-                  json_agg(gcd_component_comment) AS array_comment,
-                  json_agg(gcd_grader_id) AS array_grader_id,
-                  json_agg(gcd_graded_version) AS array_graded_version,
-                  json_agg(gcd_grade_time) AS array_grade_time,
-                  json_agg(string_mark_id) AS array_mark_id,
-
+                  json_agg(in_gcd.gc_id) as array_comp_id,
```
Comment (KevinMackenzie) [inline site/app/libraries/database/PostgresqlDatabaseQueries.php]: make all of the `as` capital (`AS`)

File: site/app/libraries/database/PostgresqlDatabaseQueries.php [hunk 7]
```diff
@@ -1390,7 +1428,6 @@ public function getGradedGradeables(array $gradeables, $users = null, $teams = n
             // Always construct an instance even if there is no data
             $auto_graded_gradeable = new AutoGradedGradeable($this->core, $graded_gradeable, $row);
             $graded_gradeable->setAutoGradedGradeable($auto_graded_gradeable);
-
```
Comment (KevinMackenzie) [inline site/app/libraries/database/PostgresqlDatabaseQueries.php]: remove this line deletion

File: site/app/libraries/database/PostgresqlDatabaseQueries.php [hunk 8]
```diff
@@ -1468,7 +1505,6 @@ public function getGradedGradeables(array $gradeables, $users = null, $teams = n
                 }
                 $ta_graded_gradeable->setGradedComponentsFromDatabase($graded_components);
             }
-
```
Comment (KevinMackenzie) [inline site/app/libraries/database/PostgresqlDatabaseQueries.php]: remove line deletion

File: migration/migrations/course/20180709170000_verifyGraders2.py [hunk 9]
```diff
@@ -0,0 +1,9 @@
+def up(config, conn, semester, course):
+    with conn.cursor() as cursor:
+        cursor.execute("ALTER TABLE ONLY gradeable_component_data DROP COLUMN IF EXISTS verifier_id")
+        cursor.execute("ALTER TABLE ONLY gradeable_component_data ADD COLUMN verifier_id character varying(255) NULL")
```
Comment (KevinMackenzie) [inline migration/migrations/course/20180709170000_verifyGraders2.py]: keep the column names consistent with the rest of the columns in that table (i.e. `gcd_verifier_id`)

File: migration/data/course_tables.sql [hunk 10]
```diff
@@ -100,10 +100,10 @@ CREATE FUNCTION csv_to_numeric_gradeable(vcode text[], gradeable_id text, grader
         LOOP
           IF istext[j] THEN
           --COME BACK AND FIX: need to put in gcd_grade_time...double check to see that CSV upload still works for numeric/text
-            INSERT INTO gradeable_component_data(gc_id, gd_id, gcd_component_comment, gcd_grader_id, gcd_graded_version, gcd_grade_time) VALUES (gcids[j], gdid, line[j+1], grader_id, NULL);
+            INSERT INTO gradeable_component_data(gc_id, gd_id, gcd_component_comment, gcd_grader_id, verifier_id ,gcd_graded_version, gcd_grade_time) VALUES (gcids[j], gdid, line[j+1], grader_id, NULL, NULL);
```
Comment (KevinMackenzie) [inline migration/data/course_tables.sql]: Fix name `verifier_id` here...

File: migration/data/course_tables.sql [hunk 11]
```diff
@@ -100,10 +100,10 @@ CREATE FUNCTION csv_to_numeric_gradeable(vcode text[], gradeable_id text, grader
...
           ELSE
             score := CAST(line[j+1] AS NUMERIC);
-            INSERT INTO gradeable_component_data(gc_id, gd_id, gcd_score, gcd_grader_id, gcd_graded_version, gcd_grade_time) VALUES (gcids[j], gdid, score, grader_id, NULL);
+            INSERT INTO gradeable_component_data(gc_id, gd_id, gcd_score, gcd_grader_id, verifier_id, gcd_graded_version, gcd_grade_time) VALUES (gcids[j], gdid, score, grader_id, NULL, NULL);
```
Comment (KevinMackenzie) [inline migration/data/course_tables.sql]: ...and here

File: migration/data/course_tables.sql [hunk 12]
```diff
@@ -257,6 +257,7 @@ CREATE TABLE gradeable_component_data (
     gcd_score numeric NOT NULL,
     gcd_component_comment character varying NOT NULL,
     gcd_grader_id character varying(255) NOT NULL,
+    verifier_id character varying(255) NULL,
```
Comment (KevinMackenzie) [inline migration/data/course_tables.sql]: ...and here

File: migration/data/course_tables.sql [hunk 13]
```diff
@@ -837,6 +838,13 @@ ALTER TABLE ONLY gradeable_component_data
 ALTER TABLE ONLY gradeable_component_data
   ADD CONSTRAINT gradeable_component_data_gcd_grader_id_fkey FOREIGN KEY (gcd_grader_id) REFERENCES users(user_id) ON UPDATE CASCADE;
 
+--
+-- Name: gradeable_component_data_verifier_id_fkey; Type: FK CONSTRAINT; Schema: public; Owner: -
+--
+
+ALTER TABLE ONLY gradeable_component_data
+  ADD CONSTRAINT gradeable_component_data_verifier_id_fkey FOREIGN KEY (verifier_id) REFERENCES users(user_id) ON UPDATE CASCADE;
```
Comment (KevinMackenzie) [inline migration/data/course_tables.sql]: ...and here

File: site/app/controllers/grading/SimpleGraderController.php [hunk 14]
```diff
@@ -174,7 +174,6 @@ public function grade($action) {
                     . $user->getLastName()
                     . ' <' . $user->getId() . '>'];
         }, $students));
-
```
Comment (KevinMackenzie) [inline site/app/controllers/grading/SimpleGraderController.php]: remove whitespace change in otherwise unchanged file

**Comment by jchulton** [comment]
So should we go from tabs to spaces (not a great pull request for it, but it work break anything)?

File: site/app/libraries/database/PostgresqlDatabaseQueries.php (hunk 7 above)
Comment (KevinMackenzie) [reply site/app/libraries/database/PostgresqlDatabaseQueries.php]: Still do remove this line deletion and the other one a hundred or so lines down.   Removing these blank lines reduces readability.

File: site/app/models/Config.php [hunk 15]
```diff
@@ -167,6 +168,8 @@ class Config extends AbstractModel {
     protected $forum_enabled;
     /** @property @var bool */
     protected $regrade_enabled;
+    /** @property @var bool */
+    protected $manual_grading_enabled;
```
Comment (KevinMackenzie) [inline site/app/models/Config.php]: Why is this a course-wide setting?

File: site/app/models/GradeableComponent.php [hunk 16]
```diff
@@ -120,6 +127,12 @@ public function __construct(Core $core, $details=array()) {
         if (isset($details['gcd_score']) && $details['gcd_score'] !== null) {
             $this->has_grade = true;
             $this->grader = $details['gcd_grader'];
+            if(isset($details['gcd_verifier'])){
+                $this->verifier=$details['gcd_verifier'];
+            }
+            else{
+                $this->verifier=null;
+            }
```
Comment (KevinMackenzie) [inline site/app/models/GradeableComponent.php]: use the `??` operator here (`blah = array['key'] ?? fallback_value`)

File: site/app/models/Config.php (hunk 15 above)
Comment (jchulton) [reply site/app/models/Config.php]: When the manual_grading_enabled option is unselected, the verify button is hidden (so no more verification). I vaguely remember Barb mentioning turning on and off verifying: I'll run it by her when she gets back. Also I renamed "manual grading enabled" to "verify enabled"

File: site/app/templates/grading/electronic/Status.twig [hunk 17]
```diff
@@ -39,11 +45,25 @@
...
+                            <th>Graded</th>
+                            <th>Verified</th>
+                            <th>Double-Graded</th>
+                        </tr>
                     <div style="margin-left: 20px">
+                        {% set counter = 0 %}
                         {% for key, section in sections %}
-                            Section {{ key }}: {{ section.graded }} / {{ section.total }} ({{ section.percentage }}%)<br />
+                            <tr>
+                                <td style="padding: 0 10px 0 10px;">{{ key }}</td>
```
[the remark this answers is not in the document]
Comment (CouleeApps) [reply site/app/templates/grading/electronic/Status.twig]: ^

File: site/app/libraries/database/DatabaseQueries.php [hunk 18]
```diff
@@ -1287,10 +1362,18 @@ public function updateGradeableData(Gradeable $gradeable) {
      * @param GradeableComponent $component
      */
     public function insertGradeableComponentData($gd_id, GradeableComponent $component) {
```
[the remark this answers is not in the document]
Comment (jchulton) [reply site/app/libraries/database/DatabaseQueries.php]: I need the if statement because if $component->getVerifier() is null, $component->getVerifier()->getId() will throw an error (calling getId() on a null)

[the remark this answers is not in the document]
Comment (KevinMackenzie) [reply site/app/libraries/database/DatabaseQueries.php]: Ok, I see that, but you can still achieve the same query without duplicating everything.

**Comment by bmcutler** [comment]
will be a new PR
